modules/core/src/test/java/org/apache/ignite/internal/processors/cache/consistency/AbstractCacheConsistencyTest.java

```diff
@@ -0,0 +1,336 @@
+import org.junit.Before;
+
+import static org.apache.ignite.cache.CacheAtomicityMode.TRANSACTIONAL;
+import static org.apache.ignite.cache.CacheMode.PARTITIONED;
+import static org.apache.ignite.cache.CacheWriteSynchronizationMode.FULL_SYNC;
+import static org.apache.ignite.events.EventType.EVT_CONSISTENCY_VIOLATION;
+
+/**
+ *
+ */
+public abstract class AbstractCacheConsistencyTest extends GridCommonAbstractTest {
+    /** Events. */
+    private static ConcurrentLinkedDeque<CacheConsistencyViolationEvent> evtDeq = new ConcurrentLinkedDeque<>();
+
+    /** Key. */
+    protected static int iterableKey;
+
+    /** Is client flag. */
+    protected boolean client;
+
+    /** Backups count. */
+    protected Integer backupsCount() {
+        return 3;
+    }
+
+    /** Cache mode. */
+    protected CacheMode cacheMode() {
+        return PARTITIONED;
+    }
+
+    /** Atomicy mode. */
+    protected CacheAtomicityMode atomicyMode() {
+        return TRANSACTIONAL;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTestsStarted() throws Exception {
+        super.beforeTestsStarted();
+
+        Ignite ignite = startGrids(7); // Server nodes.
+
+        grid(0).getOrCreateCache(cacheConfiguration());
+
+        client = true;
+
+        startGrid(G.allGrids().size() + 1); // Client node 1.
+        startGrid(G.allGrids().size() + 1); // Client node 2.
+
+        final IgniteEvents evts = ignite.events();
+
+        evts.remoteListen(null,
+            (IgnitePredicate<Event>)e -> {
+                assert e instanceof CacheConsistencyViolationEvent;
+
+                evtDeq.add((CacheConsistencyViolationEvent)e);
+
+                return true;
+            },
+            EVT_CONSISTENCY_VIOLATION);
+
+        awaitPartitionMapExchange();
+    }
+
+    /**
+     *
+     */
+    @Before
+    public void before() {
+        evtDeq.clear();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTestsStopped() throws Exception {
+        super.afterTestsStopped();
+
+        log.info("Checked " + iterableKey + " keys");
+
+        stopAllGrids();
+    }
+
+    /**
+     *
+     */
+    protected CacheConfiguration<Integer, Integer> cacheConfiguration() {
+        CacheConfiguration<Integer, Integer> cfg = new CacheConfiguration<>(DEFAULT_CACHE_NAME);
+
+        cfg.setWriteSynchronizationMode(FULL_SYNC);
+        cfg.setCacheMode(cacheMode());
+        cfg.setAtomicityMode(atomicyMode());
+        cfg.setBackups(backupsCount());
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        cfg.setClientMode(client);
+
+        return cfg;
+    }
+
+    /**
+     *
+     */
+    protected static void checkEvent(ReadRepairData data) {
+        Map<Object, Object> fixed = new HashMap<>();
+
+        while (!evtDeq.isEmpty()) {
```

Comment:
Can be replaced with a oneliner `evtDeq.forEach(e -> fixed.putAll(e.getFixedEntries()));`

Reply:
forEach is not causes removes AFAIK.
Code like 

        evtDeq.forEach(e -> {
            CacheConsistencyViolationEvent evt = evtDeq.remove();

            fixed.putAll(evt.getFixedEntries()); // Optimistic and read commited transactions produce per key fixes.
        });

looks odd to me, since it used external map API inside the forEach.